### ISSUE TITLE
Audit use of __ARM_NEON__ define

### ISSUE THREAD

**Issue by malfet** [pytorch/pytorch#61046]
It looks like __ARM_NEON__ define sometimes used to distinguish between ARMv8 and ARMv7 builds, for example here:
https://github.com/pytorch/pytorch/blob/0b8a7daa2a7e765270b85234b868b5897a9822bb/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L2482

But this is not the case, as __ARM_NEON__ is defined when compiling for aarch64 using clang compiler:
```
 % clang -dM -E - < /dev/null|grep -i ARM_NEON
#define __ARM_NEON 1
#define __ARM_NEON_FP 0xE
#define __ARM_NEON__ 1
```

Also see this issue, which claims that `__ARM_NEON__` is legacy define and `__ARM_NEON` should be used instead(which on the other hand will always be defined when compiling for 64-bit ARM):
https://github.com/openwall/john/issues/1998

**Comment by vkuzo** [comment]
@malfet , safe to close this?

**Comment by malfet** [comment]
@vkuzo, no, as problem is still present in codebase, and sometimes it is assumed it is defined on both ARMv7 and v8, but sometimes it used to distinguish between 32-bit and 64-bit ARM builds.

**Comment by andrewor14** [comment]
Hi @malfet, any updates on this issue? It's been two years since the last comment. Is it still an issue today?
